compute_endpoint/globus_compute_endpoint/pam/__init__.py

```diff
@@ -0,0 +1,338 @@
+from __future__ import annotations
+
+import typing as t
+from ctypes import CDLL, CFUNCTYPE, POINTER, Structure, byref, c_char_p, c_int, c_void_p
+from ctypes.util import find_library
+from enum import Enum
+
+pam_lib_name = find_library("pam")
+if pam_lib_name is None:
+    msg = (
+        "Failed to find `pam` shared library\n"
+        "\n  Hints:"
+        "\n    - install libpam.so? (via yum, apt, etc.)"
+        "\n    - ldconfig -p"
+        "\n    - LD_PRELOAD"
+    )
+    raise ImportError(msg)
+
+libpam = CDLL(pam_lib_name)
+
+
+# Constants (interpreted here as Enums) as defined in libpam; see _pam_types.h
+PAM_SUCCESS = 0
+
+
+class PamReturnEnum(Enum):
+    PAM_SUCCESS = PAM_SUCCESS
+
+    PAM_OPEN_ERR = 1  # dlopen() failure when dynamically loading a service module
+    PAM_SYMBOL_ERR = 2
+    PAM_SERVICE_ERR = 3
+    PAM_SYSTEM_ERR = 4
+    PAM_BUF_ERR = 5
+    PAM_PERM_DENIED = 6
+    PAM_AUTH_ERR = 7
+    PAM_CRED_INSUFFICIENT = 8
+    PAM_AUTHINFO_UNAVAIL = 9
+    PAM_USER_UNKNOWN = 10
+    PAM_MAXTRIES = 11
+    PAM_NEW_AUTHTOK_REQD = 12
+    PAM_ACCT_EXPIRED = 13
+    PAM_SESSION_ERR = 14
+    PAM_CRED_UNAVAIL = 15
+    PAM_CRED_EXPIRED = 16
+    PAM_CRED_ERR = 17
+    PAM_NO_MODULE_DATA = 18
+    PAM_CONV_ERR = 19
+    PAM_AUTHTOK_ERR = 20
+    PAM_AUTHTOK_RECOVERY_ERR = 21
+    PAM_AUTHTOK_LOCK_BUSY = 22
+    PAM_AUTHTOK_DISABLE_AGING = 23
+    PAM_TRY_AGAIN = 24
+    PAM_IGNORE = 25
+    PAM_ABORT = 26
+    PAM_AUTHTOK_EXPIRED = 27
+    PAM_MODULE_UNKNOWN = 28
+    PAM_BAD_ITEM = 29
+    PAM_CONV_AGAIN = 30
+    PAM_INCOMPLETE = 31
+
+
+class PamPrompt(Enum):
+    # message styles
+    PAM_PROMPT_ECHO_OFF = 1
+    PAM_PROMPT_ECHO_ON = 2
+    PAM_ERROR_MSG = 3
+    PAM_TEXT_INFO = 4
+
+
+class PamAuthenticate(Enum):
+    PAM_DISALLOW_NULL_AUTHTOK = 0x1
+    PAM_SILENT = 0x8000
+
+
+class PamCred(Enum):
+    PAM_ESTABLISH_CRED = 0x2
```

Comment:
I'm trying to decide how much we should care, particularly since we don't directly support Solaris, but `pamlib` does include a compat module with different values [for these constants for Solaris](https://github.com/linux-pam/linux-pam/blob/940747f88c16e029b69a74e80a2e94f65cb3e628/libpam/include/security/_pam_compat.h#L38).  While this may not be an immediate concern, but since there is precedence I would feel a little better if we were able to pull these from the headers directly.  One way to do that would be to include a c module that gets compiled against the headers installed on that machine, but there may be a more elegant way.

Reply:
As discussed, the only avenue either of us can think of to get access to the `#define`d integers is to force a preprocessor against the PAM headers in a module we include.  Not really what we want to do, but ... if the values can change&nbsp;&mdash;&nbsp;and precedent suggests they can&nbsp;&mdash;&nbsp;then we should do it.

What we have is safe for round 1, given that we only support Linux, but I will create a story for us to prioritize and get to this task.  Leaving this thread unresolved until I complete that task.

Reply:
Ticket created: sc-37696.